### ISSUE TITLE
problems with running topic-collision

### ISSUE THREAD

**Issue by paschk31** [ComputationalRadiationPhysics/picongpu#3461]
topic-collision: https://github.com/ComputationalRadiationPhysics/picongpu/pull/3416  (@pordyna)

Simulations with collisions (adjusted in collision.param) either run only up to ~20% for low laser energies or up to 100% without output for high laser energy. For the low energies i get no error, just that the time ran out (i set it to 20hrs). The same simulations without collisions (turned off in collision.param) run wihtout a problem. 

I tried making the box smaller (to a half), reducing ppc (to a half, 32) and used more gpus (8) but its gets just a tiny bit better (runs up to ~30%) for some and for other there is no change. The high energies run without output for every adjustment.

here are the input setups (cases in cmakeFlags): 
[a5_tp4_n31_input.zip](https://github.com/ComputationalRadiationPhysics/picongpu/files/5658618/a5_tp4_n31_input.zip)
[a5_tp40_n31_input.zip](https://github.com/ComputationalRadiationPhysics/picongpu/files/5658622/a5_tp40_n31_input.zip)

(the densities as shown in cmakeFlags are incorrect, real values are a sixth of that because i messed up the density ratio)

here are the energy diagrams:

**collisions turned on:**
100% = 200000 iterations

![collision_a5_tp4](https://user-images.githubusercontent.com/65019161/101472184-87943e00-3948-11eb-9794-7ce11dab367e.png)
![collision_a5_tp40](https://user-images.githubusercontent.com/65019161/101472227-9aa70e00-3948-11eb-8994-e3b63948ba87.png)
![collision_a100_tp4](https://user-images.githubusercontent.com/65019161/101472235-a0045880-3948-11eb-8903-f87bc6d2d63a.png)
![collision_a100_tp40](https://user-images.githubusercontent.com/65019161/101472239-a1ce1c00-3948-11eb-8508-30e18d34aefa.png)

﻿﻿﻿﻿﻿
**collisions turned off:**
![nocolliaion_a5_tp4](https://user-images.githubusercontent.com/65019161/101472309-bd392700-3948-11eb-9917-e675250e7981.png)
![nocollision_a5_tp40](https://user-images.githubusercontent.com/65019161/101472317-c1654480-3948-11eb-8270-bf8cb18da3d5.png)
![nocollision_a100_tp4](https://user-images.githubusercontent.com/65019161/101472326-c4603500-3948-11eb-8162-984ee4e20490.png)
![nocollision_a100_tp40](https://user-images.githubusercontent.com/65019161/101472334-c6c28f00-3948-11eb-9982-246901bb1c3a.png)

﻿﻿﻿

**Comment by PrometheusPi** [comment]
@paschk31 at 200k iterations and a walltime of 20h your simulations require a 360ms per iteration speed. This is already very fast without collisions but IO included (of course it highly depends on your setup). 

Most likely you run into the walltime, because with collisions included, one iteration requires more than a second. Cold you confirm this via the `stdout` output? 

To overcome the issue, please activate [checkpoints](https://picongpu.readthedocs.io/en/latest/usage/plugins/checkpoint.html) and restart from the checkpoint after you reached the walltime.

**Comment by sbastrakov** [comment]
Hello @paschk31 ,

The data missing on plots can be caused by two reasons:

- data is indeed not part of the output
- data is not valid, e.g. NaN or maybe not in valid range depending on how you plot it

I would imagine the data being invalid (e.g. due to a bug in collisions or otherwise) is actually more likely than somehow some data being in the output and other not. And that is easy to check by manually looking at the respective output files. Could you please check it? Or somehow share it in case you are unsure how to do so.

What @PrometheusPi wrote is also a very valid concern, but I think it's not what happened yet.

**Comment by paschk31** [comment]
@PrometheusPi okay i will try checkpoints, thanks! 
The iterations take quite a normal time at the beginning, 0% to 10% takes like 10 minutes, after that the same range takes up to 7 hrs. After the bigdata maintenace i will add a `stdout` to the issue. 
Nevertheless the checkpoints are not possible for the high energy range (where the simulations run up to 100%, so it is not clear where to checkpoint). 

@sbastrakov yes im not sure how to check the validity of the output, i will add these files as well. thanks!

**Comment by PrometheusPi** [comment]
@paschk31 To quickly check the actual run time/number of iterations, could you please post a `stdout` file of one of your simulations with collisions.

**Comment by psychocoderHPC** [comment]
> Hello @paschk31 ,
> 
> The data missing on plots can be caused by two reasons:
> 
> * data is indeed not part of the output
> * data is not valid, e.g. NaN or maybe not in valid range depending on how you plot it
> 
> I would imagine the data being invalid (e.g. due to a bug in collisions or somewhere else) is actually more likely than somehow some data being in the output and other not. And that is easy to check by manually looking at the respective output files. Could you please check it? Or somehow share it in case you are unsure how to do so.
> 
> What @PrometheusPi wrote is also a very valid concern, but I think it's not what happened yet.

As @sbastrakov said we should check first if there are NANs in the simulation.
After that we need to check if the collison implementation is maybe not freeing all dynamic used memory.

After bigdate is back we need additionally to `stdout` the `stderr` file.

**Comment by paschk31** [comment]
here are the requested outputs: 
**stdout**:
[5_tp4_n31_stdout.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687827/5_tp4_n31_stdout.txt)
[5_tp40_n31_stdout.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687830/5_tp40_n31_stdout.txt)
[100_tp4_n31_stdout.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687832/100_tp4_n31_stdout.txt)
[100_tp40_n95_stdout.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687834/100_tp40_n95_stdout.txt)
**stderr**:
[5_tp4_n31_stderr.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687938/5_tp4_n31_stderr.txt)
[5_tp40_n31_stderr.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687941/5_tp40_n31_stderr.txt)
[100_tp4_n31_stderr.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687945/100_tp4_n31_stderr.txt)
[100_tp40_n95_stderr.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687946/100_tp40_n95_stderr.txt)
**simOutput/e_energy_all**:
[5_tp4_n31_e_energy_all.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687877/5_tp4_n31_e_energy_all.txt)
[5_tp40_n31_e_energy_all.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687902/5_tp40_n31_e_energy_all.txt)
[100_tp4_n31_e_energy_all.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687885/100_tp4_n31_e_energy_all.txt)
[100_tp40_n95_e_energy_all.txt](https://github.com/ComputationalRadiationPhysics/picongpu/files/5687910/100_tp40_n95_e_energy_all.txt)

@sbastrakov as you predicted there are nans in the high energy simulations

**Comment by sbastrakov** [comment]
Yes, so it means the issue is not in the output, but that the simulation somehow produces wrong results.

`Nan`s result from mathematical operations that do not have a defined meaning, nor limit in the mathematical sense, like infinity / infinity, 0 / 0, log(negative number). At the same time operations like `finite_number / 0` result in infinities, not `nan`s. And all operations where one operand is `nan` also result in `nan`, so they are speading. E.g. if during some mistake a particle momentum becomes `nan` at some point, it would then procude `nan` currents at the neighboring grid cells, and they would propagate these `nan`s around via the field solver. So to investigate you probably need to find the place where it first occurs. I guess the collision code itself is the prime suspect since without it your simulation ran fine.

**Comment by paschk31** [comment]
okay will look for the place it first occurs! thanks!

**Comment by psychocoderHPC** [comment]
I reviewed the collision PR again and found few potential sources for `NaN`'s: https://github.com/ComputationalRadiationPhysics/picongpu/pull/3416#pullrequestreview-551791987

**Comment by pordyna** [comment]
@paschk31 Can you check which commit did you use for your simulation? Let me know if you need help with it :smile:

**Comment by pordyna** [comment]
@psychocoderHPC @sbastrakov So NaNs are one thing  and I'm looking into this right now but do you have any idea why a simulation would freeze like the low energy ones here?

**Comment by pordyna** [comment]
@sbastrakov @psychocoderHPC  @franzpoeschel  A little update on this. So it appears that the source of NaNs was in precision problems for highly relativistic particles. My current solution to this is to perform binary collisions with double precision, also in a 32bit simulation. What I want to do is to check for particles velocity and have a customizable threshold for beta, from which on the double precision is used. The check would be done in single precision simulations only. 

The other thing is that this setup still freezes for smaller laser parameters. I will have to look for infinite loops again.

**Comment by pordyna** [comment]
@ComputationalRadiationPhysics/picongpu-maintainers 
I have a new update on this issue. So it looks like this line

**Comment by pordyna** [comment]
@ComputationalRadiationPhysics/picongpu-maintainers 
So I have an update on this. It looks like the simulations were freezing on this print statement https://github.com/ComputationalRadiationPhysics/picongpu/blob/52161f25a61fa9f0fd69586ee64f3a72d3e107d2/include/picongpu/particles/collision/detail/ListEntry.hpp#L75-L79
When this is commented out, the simulation fails with a memory allocation related error. For now I didn't try this on @paschk31  setups, I used a never setup from @n01r that was also hanging in a similar way. So in  that setup I'm getting  an illegal memory access error.  I also got once ` 'cudaErrorMisalignedAddress': 'misaligned address'!', terminating` but this appeared earlier in the simulation and disappeared after recompiling.    

We did try reducing the simulation size before with @paschk31  simulations. Also, according to @n01r it is very unlikely that his test simulation went out of memory.  So it looks like there is some other problem with this line  https://github.com/ComputationalRadiationPhysics/picongpu/blob/52161f25a61fa9f0fd69586ee64f3a72d3e107d2/include/picongpu/particles/collision/detail/ListEntry.hpp#L62

Could that be that we are running into a race condition here? We are requesting heap memory on all threads, maybe the compiler is not serializing this and sometimes a thread will get memory that is already allocated by another thread. @psychocoderHPC @sbastrakov do you have any ideas? We already use the same code here https://github.com/ComputationalRadiationPhysics/picongpu/blob/52161f25a61fa9f0fd69586ee64f3a72d3e107d2/include/pmacc/particles/memory/boxes/ParticlesBox.hpp#L99
but we usually request a new frame by just one thread per block, like here
https://github.com/ComputationalRadiationPhysics/picongpu/blob/52161f25a61fa9f0fd69586ee64f3a72d3e107d2/include/picongpu/particles/ParticlesInit.kernel#L173-L176 

We could maybe try using `malloc` with `onlyMaster`  and request memory for all particles and later also with `onlyMaster` assign the correct pointers according to the values in the `nppc` array.

**Comment by psychocoderHPC** [comment]
Allocating memory in parallel is fine. The printf is only executed if you get a nullptr from the malloc call. This is the case when mallocMC is running out of memory.
This sounds like your simulation is too large for the system. You should reduce the size of the simulation.

Note to create the linked list for collision we use 4 byte values. A particle in PIConGPU is mostly without extensions around 30byte, this is 13% of ar particle + some mallocMC page overhead. If the linked list is to large mallocMC must find a spot with contagious pages.

**Comment by psychocoderHPC** [comment]
> We could maybe try using `malloc` with `onlyMaster` and request memory for all particles and later also with `onlyMaster` assign the correct pointers according to the values in the `nppc` array.

No we need to create a linked list for each cell so all threads must call malloc because each thread worker is handling one cell. Allocating a big chunk of memory is slower than a parallel allocation.
